Null Pointer Exception

Short: On two out of several dozen ser videos, processing (in a thread) halts and no "processed" nor "custom" folders are prepared. The GUI idles. The NULL occurs out of view in the log, hence I didn't observe it earlier.

OS: Win11, 64 GB RAM
ver 2.9.1 dev

Steps to recreate: two .ser files consistently cause this issue. Occurs both on batch and single file processing.

Below's the entire log.

`Output directory set to F:\2025-02-01\Sun-FeHe_filter-n99--Skywatcher-OtaR62Ap400F
File 2025-02-01-1028_6-Sun-FeHe.ser
Date 2025-02-01T10:28:19.385363200
SER file contains 8697 frames
Color mode : MONO (2 bytes per pixel, depth = 16 bits)
Width: 3400, height: 224
Solar parameters Carrington rotation = 2293, B0 = -6.06°, L0 = 22.79°, P = -12.42°
Computing average image and detecting limb
Dimensions of raw image determined : 3400x8696
Processing will require approximatively 8.43GB of disk space
Detected spectral line Iron (Fe I) with binning 1 (for pixel size 2.00µm)
Starting reconstruction...
Distortion polynomial ax3 + bx2 + cx + d = 0
   - a = 3.666559105977857E-11
   - b = 1.2870386179711682E-5
   - c = -0.0428917987794106
   - d = 209.08096532048944
Memory pressure factor 1
Processing batch 1
Reconstruction performance: 26.2 MB/s
Global reconstruction performance: 22.1 MB/s
Reconstruction done. Generating images...
Black estimate 173.24
Black estimate 172.63
Black estimate 172.96
Black estimate 172.75
Black estimate 172.21
Black estimate 176.73
Black estimate 172.65
Black estimate 169.43
Black estimate 175.74
Black estimate 173.96
Black estimate 172.79
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Error while processing: me.champeau.a4j.jsolex.processing.util.ProcessingException: java.lang.NullPointerException
me.champeau.a4j.jsolex.processing.util.ProcessingException: me.champeau.a4j.jsolex.processing.util.ProcessingException: java.lang.NullPointerException
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:501)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:485)
	at java.base/java.util.concurrent.ForkJoinTask.getException(ForkJoinTask.java:555)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:573)
	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:663)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:677)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:264)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:636)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.SolexVideoProcessor.startWorkflow(SolexVideoProcessor.java:551)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.SolexVideoProcessor.generateImages(SolexVideoProcessor.java:425)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.SolexVideoProcessor.process(SolexVideoProcessor.java:238)
	at me.champeau.a4j.jsolex/me.champeau.a4j.jsolex.app.JSolEx.processSingleFile(JSolEx.java:1372)
	at me.champeau.a4j.jsolex/me.champeau.a4j.jsolex.app.JSolEx.lambda$processFileWithParams$50(JSolEx.java:969)
	at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: me.champeau.a4j.jsolex.processing.util.ProcessingException: java.lang.NullPointerException
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.util.ProcessingException.wrap(ProcessingException.java:35)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.tasks.AbstractTask.get(AbstractTask.java:65)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.workflow.ProcessingWorkflow.geometryCorrection(ProcessingWorkflow.java:169)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.workflow.ProcessingWorkflow.start(ProcessingWorkflow.java:122)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.SolexVideoProcessor.lambda$startWorkflow$19(SolexVideoProcessor.java:556)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.IntPipeline$1$1.accept(IntPipeline.java:180)
	at java.base/java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:104)
	at java.base/java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:712)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:759)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1460)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2036)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:189)
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:220)
	at java.base/java.util.Optional.of(Optional.java:113)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.util.ImageWrapper.findMetadata(ImageWrapper.java:33)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.expr.impl.AbstractFunctionImpl.findEllipseInArguments(AbstractFunctionImpl.java:66)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.expr.impl.AbstractFunctionImpl.lambda$getEllipse$0(AbstractFunctionImpl.java:53)
	at java.base/java.util.Optional.or(Optional.java:313)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.expr.impl.AbstractFunctionImpl.getEllipse(AbstractFunctionImpl.java:53)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.expr.impl.Crop.autocrop2(Crop.java:187)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.tasks.GeometryCorrector.doCall(GeometryCorrector.java:197)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.tasks.GeometryCorrector.doCall(GeometryCorrector.java:51)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.tasks.AbstractTask.call(AbstractTask.java:48)
	at me.champeau.a4j.jsolex.core/me.champeau.a4j.jsolex.processing.sun.tasks.AbstractTask.get(AbstractTask.java:63)
	... 14 more

Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Ellipse rotation angle is 89.68847022144544
Black estimate 176.50
Tilt angle: -0.19°
Black estimate 172.48
Black estimate 176.72
Black estimate 174.63
Black estimate 171.77
Black estimate 170.67
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Black estimate 174.83
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Black estimate 172.77
Black estimate 170.86
Black estimate 165.23
Black estimate 171.82
Black estimate 164.42
Black estimate 169.44
Black estimate 172.03
Black estimate 165.31
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
Black estimate 171.94
Black estimate 171.65
Detected X/Y ratio: 1.61
Detected X/Y ratio: 1.61
`


![Image](https://github.com/user-attachments/assets/3a4442b3-2b37-4359-aeb1-7480c1f91fd1)

--- a/jsolex-core/src/main/java/me/champeau/a4j/jsolex/processing/sun/tasks/GeometryCorrector.java
+++ b/jsolex-core/src/main/java/me/champeau/a4j/jsolex/processing/sun/tasks/GeometryCorrector.java
@@ -31,6 +31,7 @@ import me.champeau.a4j.jsolex.processing.util.FileBackedImage;
 import me.champeau.a4j.jsolex.processing.util.ImageWrapper;
 import me.champeau.a4j.jsolex.processing.util.ImageWrapper32;
 import me.champeau.a4j.jsolex.processing.util.MutableMap;
+import me.champeau.a4j.jsolex.processing.util.ProcessingException;
 import me.champeau.a4j.jsolex.processing.util.SolarParametersUtils;
 import me.champeau.a4j.math.Point2D;
 import me.champeau.a4j.math.image.Image;
@@ -230,21 +231,24 @@ public class GeometryCorrector extends AbstractTask<GeometryCorrector.Result> {
      * @return a circle, if detected, or null.
      */
     private Ellipse computeCorrectedCircle(double shear, double shift, double sx, double sy) {
-        var newSamples = IntStream.range(0, 32)
-            .mapToDouble(i -> i * Math.PI / 16)
-            .mapToObj(ellipse::toCartesian)
-            .map(p -> {
-                var newX = (p.x() - shift + p.y() * shear);
-                return new Point2D(newX * sx, p.y() * sy);
-            })
-            .toList();
-        Ellipse correctedEllipse = null;
-        try {
-            correctedEllipse = new EllipseRegression(newSamples).solve();
-        } catch (Exception ex) {
-            // ignore
+        int samples = 32;
+        while (samples < 1024) {
+            int finalSamples = samples;
+            var newSamples = IntStream.range(0, samples)
+                .mapToDouble(i -> 2 * i * Math.PI / finalSamples)
+                .mapToObj(ellipse::toCartesian)
+                .map(p -> {
+                    var newX = (p.x() - shift + p.y() * shear);
+                    return new Point2D(newX * sx, p.y() * sy);
+                })
+                .toList();
+            try {
+                return new EllipseRegression(newSamples).solve();
+            } catch (Exception ex) {
+                samples *= 2;
+            }
         }
-        return correctedEllipse;
+        throw new ProcessingException("Could not compute corrected ellipse");
     }
 
     public record Result(

